stellargraph/datasets/datasets.py

```diff
@@ -0,0 +1,138 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019-2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Sample datasets for stellargraph demonstrations
+"""
+
+from .dataset_loader import DatasetLoader
+from typing import Optional
+
+
+def download_all_datasets(ignore_cache: Optional[bool] = False) -> None:
+    """ download all datasets """
+    # obtain all the classes derived from DatasetLoader (note: must be direct descendants)
+    for cls in DatasetLoader.__subclasses__():
+        dataset = cls()
+        dataset.download(ignore_cache)
+
+
+class Cora(DatasetLoader):
+    def __init__(self) -> None:
+        super().__init__(
+            "Cora",
+            "cora",
+            "https://linqs-data.soe.ucsc.edu/public/lbc/cora.tgz",
+            url_archive_format="gztar",
+            expected_files=["cora.cites", "cora.content"],
+            description="The Cora dataset consists of 2708 scientific publications classified into one of seven classes. "
+            "The citation network consists of 5429 links. Each publication in the dataset is described by a 0/1-valued word vector "
+            "indicating the absence/presence of the corresponding word from the dictionary. The dictionary consists of 1433 unique words.",
+            source="https://linqs.soe.ucsc.edu/data",
+        )
+
+
+class CiteSeer(DatasetLoader):
+    def __init__(self) -> None:
+        super().__init__(
+            "CiteSeer",
+            "CiteSeer",
+            "https://linqs-data.soe.ucsc.edu/public/lbc/citeseer.tgz",
+            url_archive_format="gztar",
+            expected_files=["citeseer.cites", "citeseer.content"],
+            description="The CiteSeer dataset consists of 3312 scientific publications classified into one of six classes. "
+            "The citation network consists of 4732 links. Each publication in the dataset is described by a 0/1-valued word vector "
+            "indicating the absence/presence of the corresponding word from the dictionary. The dictionary consists of 3703 unique words.",
+            source="https://linqs.soe.ucsc.edu/data",
+        )
+
+
+class PubMedDiabetes(DatasetLoader):
+    """
+    The Pubmed Diabetes dataset consists of 19717 scientific publications from PubMed database pertaining to diabetes classified into one of three classes.
```

Comment:
It seems we're sometimes duplicating this description and URL field, and sometimes not. Maybe we could have it computed once? E.g. have `description` automatically filled out from `self.__doc__`

```
def __init__(self) -> None:
    super().__init__(
        ...
        description=self.__doc__
        ...
    )
```

Or have `__doc__` automatically filled out from the descriptions?

```python
class DatasetLoader:
    @staticmethod
    def compute_doc(cls):
        if cls.__doc__ is not None:
            raise ValueError("DatasetLoader docs should be empty before @compute_doc call")

        cls.__doc__ = f"""\
{cls.DESCRIPTION}

Further details at: {cls.SOURCE}"""
        return cls

@DatasetLoader.compute_doc
class PubMedDiabetes(DatasetLoader):
     description="...."
     source="..."

...
```

(with the latter the `description` and `source` arguments to the constructor become unnecessary, although `DatasetLoader.__init__` should probably check the properties exist.)

I'm sure there's tricks one could do with metaclasses to make this automatic too, but I would guess that's not worth it.

Reply:
This is a great idea, I've implemented slightly differently using '__init_subclass__'